iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/consensus/deletion/DeletionResourceManager.java

```diff
@@ -0,0 +1,323 @@
+  public static final String MAGIC_VERSION_V1 = "DELETION_V1";
+  private static final String REBOOT_TIME = "rebootTime";
+  private static final String MEM_TABLE_FLUSH_ORDER = "memTableFlushOrderId";
+  private static final String DELETION_FILE_NAME_PATTERN =
+      String.format(
+          "^_(?<%s>\\d+)-(?<%s>\\d+)\\%s$",
+          REBOOT_TIME, MEM_TABLE_FLUSH_ORDER, DELETION_FILE_SUFFIX);
+  private final String dataRegionId;
+  private final DeletionBuffer deletionBuffer;
+  private final File storageDir;
+  private final List<DeletionResource> deletionResources = new CopyOnWriteArrayList<>();
+  private final Lock recoverLock = new ReentrantLock();
+  private final Condition recoveryReadyCondition = recoverLock.newCondition();
+  private volatile boolean hasCompletedRecovery = false;
+
+  private DeletionResourceManager(String dataRegionId) throws IOException {
+    this.dataRegionId = dataRegionId;
+    this.storageDir =
+        new File(
+            IoTDBDescriptor.getInstance().getConfig().getPipeConsensusDeletionFileDir()
+                + File.separator
+                + dataRegionId);
+    this.deletionBuffer = new PageCacheDeletionBuffer(dataRegionId, storageDir.getAbsolutePath());
+    initAndRecover();
+    // Only after initAndRecover can we start serialize and sync new deletions.
+    this.deletionBuffer.start();
+  }
+
+  private void initAndRecover() throws IOException {
+    recoverLock.lock();
+    try {
+      if (!storageDir.exists()) {
+        // Init
+        if (!storageDir.mkdirs()) {
+          LOGGER.warn("Unable to create pipeConsensus deletion dir at {}", storageDir);
+          throw new IOException(
+              String.format("Unable to create pipeConsensus deletion dir at %s", storageDir));
+        }
+      }
+      try (Stream<Path> pathStream = Files.walk(Paths.get(storageDir.getPath()), 1)) {
+        Path[] deletionPaths =
+            pathStream
+                .filter(Files::isRegularFile)
+                .filter(path -> path.getFileName().toString().matches(DELETION_FILE_NAME_PATTERN))
+                .toArray(Path[]::new);
+
+        for (Path path : deletionPaths) {
+          try (DeletionReader deletionReader =
+              new DeletionReader(path.toFile(), this::removeDeletionResource)) {
+            deletionResources.addAll(deletionReader.readAllDeletions());
+          } catch (IOException e) {
+            LOGGER.warn(
+                "Detect file corrupted when recover DAL-{}, discard all subsequent DALs...",
+                path.getFileName());
+            break;
+          }
+        }
+        hasCompletedRecovery = true;
+        recoveryReadyCondition.signalAll();
+      }
+    } finally {
+      recoverLock.unlock();
+    }
+  }
+
+  @Override
+  public void close() {
+    LOGGER.info("Closing deletion resource manager for {}...", dataRegionId);
+    this.deletionResources.clear();
+    this.deletionBuffer.close();
+    waitUntilFlushAllDeletions();
```

Comment:
Shall the "close()" be called after "waitUntilFlushAllDeletions()"? If there are unserialized deletionResources the "wait" may wait forever.

Reply:
Good catch, fixed